src/time_integrator.f90

```diff
@@ -55,23 +63,175 @@ end function constructor
+      call self%backend%vecadd(1.5_dp * dt, self%deriv(i)%ptr, 1._dp, self%curr(i)%ptr)
+      call self%backend%vecadd(-0.5_dp * dt, self%olds(i,1)%ptr, 1._dp, self%curr(i)%ptr)
+
+      ! for startup
+      if (self%istep .eq. 1 .and. self%order .gt. 2) then
+        ! roll pointers
+        ptr => self%olds(i,2)%ptr
+        self%olds(i,2)%ptr => self%olds(i,1)%ptr
+        self%olds(i,1)%ptr => ptr
+      end if
+
+      ! update olds(1) with new derivative
+      call self%backend%vecadd(1.0_dp, self%deriv(i)%ptr, 0._dp, self%olds(i,1)%ptr)
+    end do
+
+  end subroutine adams_bashforth_2nd
+
+  subroutine adams_bashforth_3rd(self,  dt)
+    class(time_intg_t), intent(inout) :: self
+    real(dp), intent(in) :: dt
+
+    integer :: i
+    class(field_t), pointer :: ptr
+
+    do i = 1, self%nvars
+      ! update solution
+      call self%backend%vecadd(23._dp / 12._dp * dt, self%deriv(i)%ptr, 1._dp, self%curr(i)%ptr)
+      call self%backend%vecadd(-4._dp / 3._dp * dt, self%olds(i,1)%ptr, 1._dp, self%curr(i)%ptr)
+      call self%backend%vecadd(5._dp / 12._dp * dt, self%olds(i,2)%ptr, 1._dp, self%curr(i)%ptr)
+      
+      ! for startup
+      if (self%istep .eq. 2 .and. self%order .gt. 3) then
+        ! roll pointers
+        ptr => self%olds(i,3)%ptr
+        self%olds(i,3)%ptr => self%olds(i,2)%ptr
+        self%olds(i,2)%ptr => self%olds(i,1)%ptr
+        self%olds(i,1)%ptr => ptr
+      ! after startup
+      else
+        ! roll pointers
+        ptr => self%olds(i,2)%ptr
+        self%olds(i,2)%ptr => self%olds(i,1)%ptr
+        self%olds(i,1)%ptr => ptr
+      end if
+
+      ! update olds(1) with new derivative
+      call self%backend%vecadd(1.0_dp, self%deriv(i)%ptr, 0._dp, self%olds(i,1)%ptr)
+    end do
+
+  end subroutine adams_bashforth_3rd
+
+  subroutine adams_bashforth_4th(self, dt)
+    class(time_intg_t), intent(inout) :: self
+    real(dp), intent(in) :: dt
+
+    integer :: i
+    class(field_t), pointer :: ptr
+
+    do i = 1, self%nvars
+      ! update solution
+      call self%backend%vecadd(55._dp / 24._dp * dt, self%deriv(i)%ptr, 1._dp, self%curr(i)%ptr)
+      call self%backend%vecadd(-59._dp / 24._dp * dt, self%olds(i,1)%ptr, 1._dp, self%curr(i)%ptr)
+      call self%backend%vecadd(37._dp / 24._dp * dt, self%olds(i,2)%ptr, 1._dp, self%curr(i)%ptr)
+      call self%backend%vecadd(3._dp / 8._dp * dt, self%olds(i,3)%ptr, 1._dp, self%curr(i)%ptr)
+      
+      ! for startup
+      if (self%istep .eq. 3 .and. self%order .gt. 4) then
+        ! roll pointers
+        ptr => self%olds(i,4)%ptr
+        self%olds(i,4)%ptr => self%olds(i,3)%ptr
+        self%olds(i,3)%ptr => self%olds(i,2)%ptr
+        self%olds(i,2)%ptr => self%olds(i,1)%ptr
+        self%olds(i,1)%ptr => ptr
+      ! after startup
+      else
+        ! roll pointers
+        ptr => self%olds(i,3)%ptr
+        self%olds(i,3)%ptr => self%olds(i,2)%ptr
+        self%olds(i,2)%ptr => self%olds(i,1)%ptr
+        self%olds(i,1)%ptr => ptr
```

Comment:
Simply roll pointers to avoid unnecessary repeated copy. Only update the last step residuals.

Reply:
Is it possible to move rolling pointers into a function so that we call this function each time we need to roll? I think it'll look more tidy like that.